singularity-eos/eos/eos.cpp

```diff
@@ -0,0 +1,27 @@
+//------------------------------------------------------------------------------
+// © 2021-2023. Triad National Security, LLC. All rights reserved.  This
+// program was produced under U.S. Government contract 89233218CNA000001
+// for Los Alamos National Laboratory (LANL), which is operated by Triad
+// National Security, LLC for the U.S.  Department of Energy/National
+// Nuclear Security Administration. All rights in the program are
+// reserved by Triad National Security, LLC, and the U.S. Department of
+// Energy/National Nuclear Security Administration. The Government is
+// granted for itself and others acting on its behalf a nonexclusive,
+// paid-up, irrevocable worldwide license in this material to reproduce,
+// prepare derivative works, distribute copies to the public, perform
+// publicly and display publicly, and to permit others to do so.
+//------------------------------------------------------------------------------
+
+#include <singularity-eos/eos/eos.hpp>
+
+using namespace singularity;
+
+#define DECLARE_MODS(x)                                                                  \
+  template class ShiftedEOS<x>;                                                          \
+  template class ScaledEOS<x>;                                                           \
+  template class UnitSystem<x>
+
+// Just explicitly declare some templated classes to speed up compile
+// times. Just pick the most commonly used ones.
```

Comment:
What was the compile time improvement? I've heard mixed reviews and observed ambiguous results when it comes to explicit template instantiation.

Reply:
To be honest, I didn't see a difference. This was mostly me being superstitious. I also couldn't entirely eliminate a source file the way the build system is now. It did not like building a library with no translation units.

Reply:
For the size of the template parameter lists I've seen, `singularity-eos` is quite tame compared to other codes, so compile-time speedups locally probably are not noticeable. Maybe there's some benefit downstream for heavy TMP codes depending on how `singularity-eos` is used.

Reply:
@mauneyc-LANL would you recommend keeping it around then, for that reason?

Reply:
hmmm..I wouldn't recommend *not* keeping them around (?). I don't *know* there's any benefit, but there's minimal downside. I guess I don't object to them being there, but I can't affirm it's necessary or useful.

Having said that, I've gotten lost for days in macros/template compiler errors, which can be really nasty just to identify before you even get to resolving it. Granted those codes had **a lot** more of this kind of pattern, which I don't think would be an issue with this snippet.

If pressed, I would say unless we can show it helps then probably best to avoid any pitfalls.

Reply:
I would say to get rid of it unless it gives noticeable speedups in compile times.

Reply:
Removed `eos.cpp`. Library now builds as header-only in the absence of the fortran bindings. Turns out the only thing I needed to add this:
```C++
  set_target_properties(singularity-eos PROPERTIES LINKER_LANGUAGE CXX)
```
which I guarded in a check for fortran bindings (since the proper linker is obviously not exclusively C++ in that case).

Reply:
Oops I lied. That fixes local tests, but codes downstream that rely on `singularity-eos` being installed still look for `a libsingularity.a` file and they die at the linker phase. I put `eos.cpp` back in, but it's just an empty file with a long comment explaining what I did.

Reply:
Comment transcribed below:
```
// JMM: This is a dummy file to provide a source file for cmake to
// compile when built without fortran.
//
// The main reason to do this is that when we build without fortran
// bindings, the library is completely header only. However WITH
// fortran bindings, it is not.
//
// cmake supports a build mode for header only libraries by marking
// those libraries INTERFACE. Without this flag, a few things will go
// wrong:
// 1. CMake will be unable to infer a proper linker
//    so one must be specified by hand.
// 2. no dynamic library file (e.g., "*.a" will be generated,
//    but dependencies that link against it, such as another library,
//    or in our case, tests, will look for it. This will lead to
//    a failure of the downstream library at link time.
// 
// On the other hand, a library with implementation files
// CANNOT be marked INTERFACE, as otherwise source files will
// not be compiled.
//
// Unfortunately, switching the INTERFACE tag on and off in cmake
// is very cumbersome, as if a library is marked INTERFACE
// all its dependencies must ALSO be slurped in with an INTERFACE
// flag. This introduces significant branching in the cmake code
// and a lot of builer plate. For now, then, I simply include this
// empty source file.
//
// In the future, we could include centralized code here that we
// always want compiled, such as template instantiations or
// convenience functions.
```

Reply:
Would the simpler solution here just be to have a separate target for the Fortran library?

Reply:
That sounds reasonable to me. But yeah, let's leave it for a later PR.